### ISSUE TITLE
twisted.amp documentation _command and _ask order inverted

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3216]
|<img alt="proppy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| proppy reported|
|-|-|
|Trac ID|trac#3216|
|Type|defect|
|Created|2008-04-30 11:02:09Z|
|Branch|https://github.com/twisted/twisted/tree/amp-keys-doc-3216|

This documentation page: http://twistedmatrix.com/documents/current/api/twisted.protocols.amp.html

Lists amp keys in the following order:
```
C: _command: sum
C: _ask: ef639e5c892ccb54
```

However when reading raw amp data from C++:
http://proppy.aminche.com/hg/boost-asio-sample/file/f58a19591ee5/client.cpp

I received data in this order:
```
_ask
8
_command
MessageReceived
message
ping
```

Attachments:

* [twisted-amp-doc-ask-command-inverted.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ddc/ddce5d4cd1b57f31f62f971b9882cd08d46195e8/44a7a737efb979ed87d9e8810f28b5afe391b32e.patch) (409 bytes) - added by proppy on 2008-04-30 11:10:44Z - patch for amp.py docstring

<details><summary>Searchable metadata</summary>

```
trac-id__3216 3216
type__defect defect
reporter__proppy proppy
priority__normal normal
milestone__ 
branch__branches_amp_keys_doc_3216 branches/amp-keys-doc-3216
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1209553329000000 1209553329000000
changetime__1226765475000000 1226765475000000
version__None None
owner__ 
cc__thijs cc__exarkun cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

The relative order of `_ask` and `_command` is arbitrary.  Perhaps the documentation should be updated to say this.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

(In [25123]) Branching to 'amp-keys-doc-3216'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Up for review, see r25124

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

There are some mistakes.  I'm going to fix them.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [25130]) restore Divmod copyright statement

it may make sense to remove this and replace it with an entry in the LICENSE file

refs [#3216](https://github.com/twisted/twisted/issues/3216)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [25131]) NUL is not code

refs [#3216](https://github.com/twisted/twisted/issues/3216)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [25132]) expound

refs [#3216](https://github.com/twisted/twisted/issues/3216)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Okay, looks good to me now. :)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

Looks good, will merge it now.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

... why did this go from "will merge it now" to review?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

It needs a review by someone who wasn't an author.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [25346]) Merge amp-keys-doc-3216

Author: thijs, exarkun
Reviewer: therve
Fixes: [#3216](https://github.com/twisted/twisted/issues/3216)

Clarify the amp box key ordering rules in the amp module docstring.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
